File: Dsa-day232.java
class Solution {
    public int minFlips(String s) {
        int n = s.length();
        String t = s + s;

        int diff1 = 0, diff2 = 0;
        int left = 0;
        int res = Integer.MAX_VALUE;

        for (int right = 0; right < 2 * n; right++) {

            char expected1 = (right % 2 == 0) ? '0' : '1';
            char expected2 = (right % 2 == 0) ? '1' : '0';

            if (t.charAt(right) != expected1) diff1++;
            if (t.charAt(right) != expected2) diff2++;

            if (right - left + 1 > n) {
                char leftExpected1 = (left % 2 == 0) ? '0' : '1';
                char leftExpected2 = (left % 2 == 0) ? '1' : '0';

                if (t.charAt(left) != leftExpected1) diff1--;
                if (t.charAt(left) != leftExpected2) diff2--;

                left++;
            }

            if (right - left + 1 == n) {
                res = Math.min(res, Math.min(diff1, diff2));
            }
        }

        return res;
    }
}

File: Dsa-day232.py
class Solution:
    def minFlips(self, s: str) -> int:
        n = len(s)
        t = s + s

        diff1 = diff2 = 0
        left = 0
        res = float('inf')

        for right in range(2 * n):

            expected1 = '0' if right % 2 == 0 else '1'
            expected2 = '1' if right % 2 == 0 else '0'

            if t[right] != expected1:
                diff1 += 1
            if t[right] != expected2:
                diff2 += 1

            if right - left + 1 > n:
                left_expected1 = '0' if left % 2 == 0 else '1'
                left_expected2 = '1' if left % 2 == 0 else '0'

                if t[left] != left_expected1:
                    diff1 -= 1
                if t[left] != left_expected2:
                    diff2 -= 1

                left += 1

            if right - left + 1 == n:
                res = min(res, diff1, diff2)

        return res